gnome-image-installer/pages/install/gis-scribe.c

```diff
@@ -0,0 +1,820 @@
+   */
+  if (!g_output_stream_write_all (output, first_mib, buffer_size,
+                                  &w, cancellable, error)
+      || !g_input_stream_read_all (self->decompressed, first_mib, buffer_size,
+                                   &first_mib_bytes_read, cancellable, error))
+    return FALSE;
+
+  do
+    {
+      r = g_input_stream_read (self->decompressed, buffer, buffer_size,
+                               cancellable, error);
+
+      if (r < 0)
+        return FALSE;
+
+      if (!g_output_stream_write_all (output, buffer, r,
+                                      &w, cancellable, error))
+        return FALSE;
+
+      /* We lock to protect bytes_written */
+      g_mutex_lock (&self->mutex);
+      self->bytes_written += w;
+      g_mutex_unlock (&self->mutex);
+    }
+  while (r > 0);
+
+  /* Now write the first 1 MiB to disk. Unfortunately GUnixOutputStream does
+   * not implement GSeekable.
+   */
+  if (lseek (fd, 0, SEEK_SET) < 0)
+    return glnx_throw_errno_prefix (error, "can't seek to start of disk");
+
+  if (!g_output_stream_write_all (output, first_mib, first_mib_bytes_read,
+                                  &w, cancellable, error))
+    return FALSE;
+
+  g_mutex_lock (&self->mutex);
+  self->bytes_written += w;
+  g_mutex_unlock (&self->mutex);
+
+  return TRUE;
+}
+
+static gboolean
+gis_scribe_set_indeterminate_progress (gpointer data)
+{
+  GisScribe *self = GIS_SCRIBE (data);
+
+  g_mutex_lock (&self->mutex);
+  self->set_indeterminate_progress_id = 0;
+  g_mutex_unlock (&self->mutex);
+
+  self->progress = -1;
+  g_object_notify_by_pspec (G_OBJECT (self), props[PROP_PROGRESS]);
+
+  return G_SOURCE_REMOVE;
+}
+
+static gboolean
+gis_scribe_convert_to_mbr (GisScribe *self,
+                           GError   **error)
+{
+  static const char *cmd = "/usr/sbin/eos-repartition-mbr";
+  g_autoptr(GSubprocessLauncher) launcher = NULL;
+  g_autoptr(GSubprocess) process = NULL;
+
+  launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+  /* pkexec won't let us run the program if $SHELL isn't in /etc/shells,
+   * so remove it from the environment.
+   */
+  g_subprocess_launcher_unsetenv (launcher, "SHELL");
+  process = g_subprocess_launcher_spawn (launcher, error,
+                                         "pkexec", cmd, self->drive_path, NULL);
+  if (process == NULL ||
+      !g_subprocess_wait_check (process, NULL, error))
+    {
+      g_prefix_error (error, "failed to run %s %s: ",
+                      cmd, self->drive_path);
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static void
+gis_scribe_write_thread (GTask        *task,
+                         gpointer      source_object,
+                         gpointer      task_data,
+                         GCancellable *cancellable)
+{
+  GisScribe *self = GIS_SCRIBE (source_object);
+  gint fd = -1;
+  g_autoptr(GOutputStream) output = NULL;
+  gboolean ret;
+  GError *error = NULL;
```

Comment:
I disagree in this case since `error` must always be passed to `g_task_return_error()` if non-`NULL`, except in one case where it is manually cleared.